only configure swap if swap is enabled

elezar opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug
/kind regression

#### What this PR does / why we need it:

This PR fixes the startup of kubernetes on systems using cgroupv2 where swap is not enabled. This can be demonstrated on such as system using kind and any kubernetes version of 1.28.0, 1.28.1, or 1.28.2.

Since the `memory.swap.max` value is set, this causes containerd and runc to try to write a value for the cgroup causing containers to fail when starting with messages such as:
```
Sep 20 09:48:54 k8s-dra-driver-cluster-control-plane kubelet[318]: E0920 09:48:54.918262     318 pod_workers.go:1300] "Error syncing pod, skipping" err="failed to \"StartContainer\" for \"kube-apiserver\" with RunContainerError: \"failed to create co
ntainerd task: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error setting cgroup config for procHooks process: unable to set unified resource \\\"memory.swa
p.max\\\": open /sys/fs/cgroup/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-burstable.slice/kubelet-kubepods-burstable-poddfde4e625667960b6c2b494dd6946943.slice/cri-containerd-4085d63ae91fe66c6fd8ea3e7d2d20d3f7492ef51f11621e2dcbb7f47d345d0f.
scope/memory.swap.max: no such file or directory: unknown\"" pod="kube-system/kube-apiserver-k8s-dra-driver-cluster-control-plane" podUID="dfde4e625667960b6c2b494dd6946943"
```

This is independent of whether `NodeSwap` is enabled or not.

Note that the issue seemed to have been introduced in #118764 with the fix extending the logic added in #119486 to also be applicable to cgroupv2 systems with swap disabled.

<!--
#### Which issue(s) this PR fixes:

Fixes #
-->

#### Special notes for your reviewer:

Assuming:
* cgroupv2 is being used (`/sys/fs/cgroup/cgroup.controllers` is present)
* swap is disabled (`swapon -s` shows nothing)

This should be reproducible in kind with the following:
```
kind create cluster --retain --image kindest/node:v1.28.0 --name test-cluster
```
which will create a single-node kind cluster using k8s `v1.28.0`. This will cause an error when starting the control-plane node.

Running:
```
kind export logs --name test-cluster
```
Will ensure that the kubelet logs are available.  These can be inspected using
```
~$ grep memory.swap.max /tmp/769685610/test-cluster-control-plane/kubelet.log | tail -1
Sep 20 19:59:32 foo-control-plane kubelet[263]: E0920 19:59:32.930675     263 pod_workers.go:1300] "Error syncing pod, skipping" err="failed to \"StartContainer\" for \"etcd\" with RunContainerError: \"failed to create containerd task: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error setting cgroup config for procHooks process: unable to set unified resource \\\"memory.swap.max\\\": open /sys/fs/cgroup/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-burstable.slice/kubelet-kubepods-burstable-podc576cd164244f9e5e46e146c0d642304.slice/cri-containerd-8729d25356d4538d93740e37da5106fbce905d354c50af77aae373ea9d68b010.scope/memory.swap.max: no such file or directory: unknown\"" pod="kube-system/etcd-foo-control-plane" podUID="c576cd164244f9e5e46e146c0d642304"
```
Where this particular example shows the etcd container failing to start.

Furthermore, we can confirm that this error is not present in an earlier k8s version:
```
$ kind create cluster --image kindest/node:v1.27.6 --name test-cluster
Creating cluster "test-cluster" ...
 ✓ Ensuring node image (kindest/node:v1.27.6) 🖼
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-test-cluster"
You can now use your cluster with:
```


#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed a bug where containers would not start on cgroupv2 systems where swap is disabled.
```

<!--
#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:


This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>

```docs

```
-->

elezar commented:
cc @iholder101 @pacoxu

klueska commented:
/cc

klueska commented:
/triage accepted
/priority important-soon

klueska commented:
/ok-to-test

elezar commented on pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:
```diff
@@ -371,6 +371,10 @@ func (m swapConfigurationHelper) ConfigureLimitedSwap(lcr *runtimeapi.LinuxConta
 }
 
 func (m swapConfigurationHelper) ConfigureNoSwap(lcr *runtimeapi.LinuxContainerResources) {
+	if !swapControllerAvailable() {
```
As a note to the reviewers: It may be better to include this check in other code paths such as `ConfigureUnlimitedSwap`. Initially I was thinking that it would be better to do it in `configureSwap`, but would have to then handle the error cases where we are expecting to set up swap.

iholder101 replied:
How about moving it to here?

https://github.com/kubernetes/kubernetes/blob/f1703405243e6e6332d4b1a68d65bfcdad66bc05/pkg/kubelet/kuberuntime/kuberuntime_container_linux.go#L110-L110


Then, if the feature gate or the cgroup controller are missing, swap logic wouldn't be executed at all.
WDYT?

elezar replied:
Yes, I was mulling over that last night too. That might be the simplest solution for now.

iholder101 replied:
> This should only be an issue if LimitedSwap is configured though -- since on a system with swap disabled my current understanding is that max and 0 are equivilent in terms of how these are handled by both Containerd and runc.

If swap is disabled, I believe that `0`, `max`, or any other value would serve as a no-op. However, in the case where the swap controller is not present, the `memory.swap.max` file is not present and therefore cannot be written to.

> One advantage of changing the swapConfigurationHelper implementation is that it allows for errors to be logged when a requested configuration cannot be applied.

If the swap controller is not available we can't use any of swapConfigurationHelper's methods, since they all configure `memory.swap.max` which is unavailable. This is why if you implement it on the swapConfigurationHelper level you would need to check for the controller's availability within each and every one of its methods, and then handle logging and returning as a no-op in each method separately.

Since the helper is being used only in one single place, I think that checking for the controller's availability in the same place is both more clear and less error-prone, as if another method would be added in the future it won't need to explicitly check if the controller is available once again. As I see it, when the helper is being used it already assumed that swap is enabled.

Does that make sense to you?

elezar replied:
Yes, I think that makes sense. I also played around with the helper a bit last night and defined an interface:
```
type swapConfigurer interface {
   configureSwap(lcr *runtimeapi.LinuxContainerResources, swapMemory int64)
}
```
With three different implementations:
* `disabledSwapConfigurer`
* `unifiedSwapConfigurer`
* `nonUnifiedSwapConfigurer`

```
func newSwapConfigurer() swapConfigurer {
	if !swapControllerAvailable() {
		return &swapDisabledConfigurer{}
	}
	if !isCgroup2UnifiedMode() {
		return &nonUnifiedSwapConfigurer{}
	}
	return &unifiedSwapConfigurer{}
}

type swapConfigurer interface {
	configureSwap(*runtimeapi.LinuxContainerResources, int64)
}

// swapDisabledConfigurer represents a swap configurer where any operation to set the swap is a no-op.
// In cases where swap is set to a non-zero value, an error is logged.
type swapDisabledConfigurer struct{}

func (m swapDisabledConfigurer) configureSwap(lcr *runtimeapi.LinuxContainerResources, swapMemory int64) {
	if swapMemory == 0 {
		return
	}
	klog.ErrorS(fmt.Errorf("swap configuration is not supported on this system"), "swap configuration is unexpected if swap is disabled")
}

// unifiedSwapConfigurer represents a swap configurer where the swap is set in the unified cgroup v2 hierarchy.
// This sets the `memory.swap.max` cgroup value.
type unifiedSwapConfigurer struct{}

func (m unifiedSwapConfigurer) configureSwap(lcr *runtimeapi.LinuxContainerResources, swapMemory int64) {
	if lcr.Unified == nil {
		lcr.Unified = map[string]string{}
	}

	var swapString string
	if swapMemory == unlimitedSwap {
		swapString = "max"
	} else {
		swapString = fmt.Sprintf("%d", swapMemory)
	}
	lcr.Unified[cm.Cgroup2MaxSwapFilename] = swapString
}

// nonUnifiedSwapConfigurer represents a swap configurer where the swap is set in the non-unified cgroup v1 hierarchy.
// Unless swap is being disabled (or set ot max), an error is logged.
type nonUnifiedSwapConfigurer struct{}

func (m nonUnifiedSwapConfigurer) configureSwap(lcr *runtimeapi.LinuxContainerResources, swapMemory int64) {
	if swapMemory == 0 || swapMemory == unlimitedSwap {
		// memorySwapLimit = total permitted memory+swap; if equal to memory limit, => 0 swap above memory limit
		// Some swapping is still possible.
		// Note that if memory limit is 0, memory swap limit is ignored.
		lcr.MemorySwapLimitInBytes = lcr.MemoryLimitInBytes
	}
	klog.ErrorS(fmt.Errorf("swap configuration is not supported with cgroup v1"), "swap configuration under cgroup v1 is unexpected")
}
```

I updated the helper as follows:
```
type swapConfigurationHelper struct {
	swapConfigurer
	machineInfo cadvisorv1.MachineInfo
}

func newSwapConfigurationHelper(machineInfo cadvisorv1.MachineInfo) *swapConfigurationHelper {
	configurer := newSwapConfigurer()
	return &swapConfigurationHelper{
		swapConfigurer: configurer,
		machineInfo:    machineInfo,
	}
}
```

meaning that it delegates to the swapConfigurer to accutally apply the swap value. I did feel that this was a larger change than what is required for this fix -- but if you think it's something worth following-up on, I don't mind creating a PR with the refactor to clarify things.

iholder101 replied:
Thanks @elezar! I appreciate the work you put into it.

My initial thoughts here is that 2 of the 3 implementations for your interface can actually be used only to configure no swap access or act as a no-op. In other words, only `unifiedSwapConfigurer` has real logic into it while `disabledSwapConfigurer` and `nonUnifiedSwapConfigurer` are almost identical in the sense that they either don't do anything or configure no swap.

In the current implementation, this is being solved by two `if` statements. The one checks if the controller is available and if it isn't the swapConfigHelper isn't even being used. The other one resides within `ConfigureNoSwap` and will likely be removed in the future when cgroup v1 would be fully dropped. Moreover, these two implementations aren't going to expend in the future.

To be completely honest, I think that your current implementation looks great. It is elegant and simple to understand. My opinion is that we don't need to complicate stuff before we have a good reason to do so (see [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)).

Perhaps we can circle back to it in the future if we indeed have more cases to handle

bart0sh commented:
@elezar is it possible to cover this use case with either unit or e2e tests?
/cc

elezar commented:
> @elezar is it possible to cover this use case with either unit or e2e tests? /cc

It may be. I discovered this while trying to get our DRA driver running on Kind. Maybe @iholder101 has some ideas / pointers on where a suitable end-to-end test case could be added?

harche commented:
/cc

harche commented:
/cc @SergeyKanzhelev

elezar committed:
Use local isCgroup2UnifiedMode consistently

This change switches to using isCgroup2UnifiedMode locally to ensure
that any mocked function is also used when checking the swap controller
availability.

Signed-off-by: Evan Lezar <elezar@nvidia.com>

iholder101 commented:
Thank you @elezar! Great fix!

WDYT about prompting a warning, or at least a log entry, to let the user know that swap is not being configured? Otherwise I think that the user can be surprised and have a hard time understanding why the swap configs aren't being applied as expected.

> > @elezar is it possible to cover this use case with either unit or e2e tests? /cc
> 
> It may be. I discovered this while trying to get our DRA driver running on Kind. Maybe @iholder101 has some ideas / pointers on where a suitable end-to-end test case could be added?

w.r.t. unit tests it is entirely possible by mocking the `swapControllerAvailable()` function. I did a similar trick with the cgroup version, here's a reference:
https://github.com/kubernetes/kubernetes/blob/f1703405243e6e6332d4b1a68d65bfcdad66bc05/pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go#L1142-L1153

w.r.t. e2e tests, that's more tricky as we need to ensure that the container would be created without a swap cgroup controller. The two ways I see of doing so is either using a lane that does not have the swap controller enabled (perhaps the kind lane?) or using something like [CRI-O annotations](https://github.com/cri-o/cri-o/blob/main/pkg/annotations/annotations.go) to enforce the container to be created without a swap controller (maybe containerd have similar annotations). However this requires more exploration.

Thanks again!
Except for these minor comments, general LGTM from my side.

/cc

elezar committed:
Only configure swap if available on node

This change bypasses all logic to set swap in the linux container
resources if a swap controller is not available on node. Failing
to do so may cause errors in runc when starting a container with
a swap configuration -- even if this is set to 0.

Signed-off-by: Evan Lezar <elezar@nvidia.com>

iholder101 commented on pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go:
```diff
@@ -1151,3 +1323,14 @@ func setCgroupVersionDuringTest(version CgroupVersion) {
 }
+
+func setSwapControllerAvailableDuringTest(available bool) func() {
+	original := swapControllerAvailable
+	swapControllerAvailable = func() bool {
+		return available
+	}
+
+	return func() {
+		swapControllerAvailable = original
+	}
```
Isn't doing the following enough?
```go
swapControllerAvailable = func() bool {
	return available
}
```

I'm pretty sure that all test cases run independently, so no need to revert to the original function value

elezar replied:
Yes, it may be enough. I was thinking that there may be some issue in the case of failures, for example. If you feel strongly about this, I can revert.

iholder101 replied:
I don't feel strongly about it, but if code can be simplified for "free" it's always a win :)
Feel free to leave it as is, we can also circle back to it in future follow-ups.

Thank you!

elezar replied:
OK. Leaving as is for now. My motivation for doing this was to align with how the feature gates are handled in test.

iholder101 commented on pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:
```diff
@@ -320,7 +320,7 @@ func swapControllerAvailable() bool {
 	swapControllerAvailabilityOnce.Do(func() {
 		const warn = "Failed to detect the availability of the swap controller, assuming not available"
 		p := "/sys/fs/cgroup/memory/memory.memsw.limit_in_bytes"
-		if libcontainercgroups.IsCgroup2UnifiedMode() {
+		if isCgroup2UnifiedMode() {
```
:+1:!

klueska commented:
@iholder101 who was the original approver of the PR that merged this change?
I am happy to approve this PR once you give your LGTM, I just want to make sure we keep the original approver in the loop as well.

iholder101 commented:
> @iholder101 who was the original approver of the PR that merged this change? I am happy to approve this PR once you give your LGTM, I just want to make sure we keep the original approver in the loop as well.

Thank you @klueska.
The original approver was @SergeyKanzhelev. Thanks for your comment, I agree it's important to loop him in as well.

klueska commented:
Hmm. @SergeyKanzhelev doesn't actually have the required approval rights (though maybe he is the most familiar with the work and did it by-proxy). Was there someone else from the sig-node-approces list that approved it: https://github.com/kubernetes/kubernetes/blob/master/OWNERS_ALIASES#L215

iholder101 commented:
> Hmm. @SergeyKanzhelev doesn't actually have the required approval rights (though maybe he is the most familiar with the work and did it by-proxy). Was there someone else from this `sig-node-approver` list that approved it: https://github.com/kubernetes/kubernetes/blob/master/OWNERS_ALIASES#L215

Yes, was lgtmed by @mrunalp

elezar commented:
> @elezar is it possible to cover this use case with either unit or e2e tests? /cc

@bart0sh I have added unit tests for the swap disabled case.

elezar commented:
Sorry for the back and forth on this. I have my development environment set up a bit better now. I also rebased off `master` to address any possible unrelated test failures.

iholder101 commented:
> Sorry for the back and forth on this. I have my development environment set up a bit better now. I also rebased off `master` to address any possible unrelated test failures.

No worries at all!

From my side, the only missing part is a warning log entry as mentioned in the [comment above](https://github.com/kubernetes/kubernetes/pull/120784#issuecomment-1729196407):
> WDYT about prompting a warning, or at least a log entry, to let the user know that swap is not being configured? Otherwise I think that the user can be surprised and have a hard time understanding why the swap configs aren't being applied as expected.

After that's there and tests pass I'm ready to LGTM.

Thank you again for your work!

elezar commented:
> From my side, the only missing part is a warning log entry as mentioned in the https://github.com/kubernetes/kubernetes/pull/120784#issuecomment-1729196407:

Ah yes. Let me add that. Do you have any suggestions on wording for the warning?

iholder101 commented:
> > From my side, the only missing part is a warning log entry as mentioned in the [#120784 (comment)](https://github.com/kubernetes/kubernetes/pull/120784#issuecomment-1729196407):
> 
> Ah yes. Let me add that. Do you have any suggestions on wording for the warning?

Ehum.. Maybe something like
```
cannot configure swap for container %s: swap cgroup controller not present
```

Also I think that you should prompt it only if the NodeSwap feature gate is enabled. There's maybe a need to re-order the condition checks in code, as currently the feature gate ennoblement is being checked only if the controller is available.

iholder101 commented:
> > From my side, the only missing part is a warning log entry as mentioned in the [#120784 (comment)](https://github.com/kubernetes/kubernetes/pull/120784#issuecomment-1729196407):
> 
> Ah yes. Let me add that. Do you have any suggestions on wording for the warning?
> 
> How about:
> 
> ```
> diff --git a/pkg/kubelet/kuberuntime/kuberuntime_container_linux.go b/pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
> index dfe1f7465be..e0e094341be 100644
> --- a/pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
> +++ b/pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
> @@ -121,6 +121,8 @@ func (m *kubeGenericRuntimeManager) generateLinuxContainerResources(pod *v1.Pod,
>  		} else {
>  			swapConfigurationHelper.ConfigureNoSwap(lcr)
>  		}
> +	} else if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
> +		klog.WarningS("NodeSwap feature is enabled, but swap controller is not available; ignoring NodeSwap feature")
>  	}
>  
>  	// Set memory.min and memory.high to enforce MemoryQoS
> ```

Looks great :+1:

elezar commented:
/retest

elezar commented:
/retest

elezar commented:
/retest

iholder101 commented:
Thank you @elezar!
/lgtm

iholder101 commented:
@mrunalp @klueska PTAL

klueska commented on pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:
```diff
@@ -107,17 +107,22 @@ func (m *kubeGenericRuntimeManager) generateLinuxContainerResources(pod *v1.Pod,
+	// swap is only configured if a swap controller is available.
+	if swapControllerAvailable() {
+		if swapConfigurationHelper := newSwapConfigurationHelper(*m.machineInfo); utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
+			// NOTE(ehashman): Behaviour is defined in the opencontainers runtime spec:
+			// https://github.com/opencontainers/runtime-spec/blob/1c3f411f041711bbeecf35ff7e93461ea6789220/config-linux.md#memory
+			switch m.memorySwapBehavior {
+			case kubelettypes.LimitedSwap:
+				swapConfigurationHelper.ConfigureLimitedSwap(lcr, pod, container)
+			default:
+				swapConfigurationHelper.ConfigureUnlimitedSwap(lcr)
+			}
+		} else {
+			swapConfigurationHelper.ConfigureNoSwap(lcr)
 		}
-	} else {
-		swapConfigurationHelper.ConfigureNoSwap(lcr)
+	} else if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
+		klog.ErrorS(errors.New("no cgroup swap controller present"), "ignoring NodeSwap feature", "pod", klog.KObj(pod), "containerName", container.Name)
 	}
```
Is there some way to cleanup these if/elses to read more nicely. In partifular, this is harad to look at:
```
if swapConfigurationHelper := newSwapConfigurationHelper(*m.machineInfo); utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
```

What about:
```
	if !swapControllerAvailable() && !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
		// Nothing to do
	}
	if !swapControllerAvailable() && utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
		klog.ErrorS(errors.New("no cgroup swap controller present"), "ignoring NodeSwap feature", "pod", klog.KObj(pod), "containerName", container.Name)
	}
	if swapControllerAvailable() && !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
		swapConfigurationHelper := newSwapConfigurationHelper(*m.machineInfo);
		swapConfigurationHelper.ConfigureNoSwap(lcr)
	}
	if swapControllerAvailable() && utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
		swapConfigurationHelper := newSwapConfigurationHelper(*m.machineInfo);
		// NOTE(ehashman): Behaviour is defined in the opencontainers runtime spec:
		// https://github.com/opencontainers/runtime-spec/blob/1c3f411f041711bbeecf35ff7e93461ea6789220/config-linux.md#memory
		switch m.memorySwapBehavior {
		case kubelettypes.LimitedSwap:
			swapConfigurationHelper.ConfigureLimitedSwap(lcr, pod, container)
		default:
			swapConfigurationHelper.ConfigureUnlimitedSwap(lcr)
		}
	}
```

elezar replied:
I was wondering about the nesting too. What about adding an `addSwapResources` method that we call here instead.

```
func (m *kubeGenericRuntimeManager) addSwapResources(...) {
	if !swapControllerAvailable() && !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
		return
	}
	if !swapControllerAvailable() {
		klog.ErrorS(errors.New("no cgroup swap controller present"), "ignoring NodeSwap feature", "pod", klog.KObj(pod), "containerName", container.Name)
		return
	}
	swapConfigurationHelper := newSwapConfigurationHelper(*m.machineInfo)
	if !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
		swapConfigurationHelper.ConfigureNoSwap(lcr)
		return
	}
	// NOTE(ehashman): Behaviour is defined in the opencontainers runtime spec:
	// https://github.com/opencontainers/runtime-spec/blob/1c3f411f041711bbeecf35ff7e93461ea6789220/config-linux.md#memory
	swapConfigurationHelper := newSwapConfigurationHelper(*m.machineInfo)
	switch m.memorySwapBehavior {
	case kubelettypes.LimitedSwap:
		swapConfigurationHelper.ConfigureLimitedSwap(lcr, pod, container)
	default:
		swapConfigurationHelper.ConfigureUnlimitedSwap(lcr)
	}
}
```

klueska replied:
definitely seems cleaner to me

elezar replied:
@klueska I reworked this a bit with quick-returns where applicable. If you really want me to check the feature gate on all if statements, I can do so.

klueska reviewed (commented):
Minor comment otherwise LGTM

sftim commented on pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:
```diff
@@ -107,17 +107,22 @@ func (m *kubeGenericRuntimeManager) generateLinuxContainerResources(pod *v1.Pod,
+	// swap is only configured if a swap controller is available.
+	if swapControllerAvailable() {
+		if swapConfigurationHelper := newSwapConfigurationHelper(*m.machineInfo); utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
+			// NOTE(ehashman): Behaviour is defined in the opencontainers runtime spec:
```
```suggestion
			// NOTE(ehashman): Behavior is defined in the opencontainers runtime spec:
```

elezar replied:
Done.

sftim commented on pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go:
(nit) In tests, please write “QoS” not “Qos”

elezar replied:
Updated.

sftim commented on pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:
```diff
@@ -107,17 +107,22 @@ func (m *kubeGenericRuntimeManager) generateLinuxContainerResources(pod *v1.Pod,
-	} else {
-		swapConfigurationHelper.ConfigureNoSwap(lcr)
+	} else if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
+		klog.ErrorS(errors.New("no cgroup swap controller present"), "ignoring NodeSwap feature", "pod", klog.KObj(pod), "containerName", container.Name)
```
Try to write the error logging as if the feature is already stable. The value of the feature gate doesn't seem very relevant here.

elezar replied:
Thanks @sftim. I elected to always log (at `InfoS`) that we're not configuring swap but to include the `SwapBehavior` value as well. This should be enough to give users a signal that its being skipped without making the logic in the function overly complex. Hope this is sufficient from your side.

sftim reviewed (commented):
Some feedback - hope it helps

elezar commented on pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:
```diff
@@ -107,17 +107,22 @@ func (m *kubeGenericRuntimeManager) generateLinuxContainerResources(pod *v1.Pod,
-			swapConfigurationHelper.ConfigureLimitedSwap(lcr, pod, container)
-		default:
-			swapConfigurationHelper.ConfigureUnlimitedSwap(lcr)
+	// swap is only configured if a swap controller is available.
```
Thanks @sftim. Should be addressed in latest revision. Note that I factored out this logic into a method, so the comment is now included as part of its docstring.

klueska commented:
@elezar thanks for tracking down this bug and fixing it!
Once this change gets merged we should back-port it to for the next 1.28 patch release.

/lgtm
/approve
